plugins/modules/source_control/gitlab/gitlab_project_variable.py

```diff
@@ -162,104 +229,91 @@ def list_all_project_variables(self):
         }
-        if environment_scope is not None:
-            var["environment_scope"] = environment_scope
         return self.project.variables.create(var)
 
-    def update_variable(self, key, var, value, masked, protected, variable_type, environment_scope):
-        if (var.value == value and var.protected == protected and var.masked == masked
-                and var.variable_type == variable_type
-                and (var.environment_scope == environment_scope or environment_scope is None)):
-            return False
-
+    def update_variable(self, var_obj):
         if self._module.check_mode:
             return True
-
-        if (var.protected == protected and var.masked == masked
-                and var.variable_type == variable_type
-                and (var.environment_scope == environment_scope or environment_scope is None)):
-            var.value = value
-            var.save()
-            return True
-
-        self.delete_variable(key)
-        self.create_variable(key, value, masked, protected, variable_type, environment_scope)
+        self.delete_variable(var_obj)
+        self.create_variable(var_obj)
         return True
 
-    def delete_variable(self, key):
+    def delete_variable(self, var_obj):
         if self._module.check_mode:
-            return
-        return self.project.variables.delete(key)
+            return True
+        self.project.variables.delete(var_obj.get('key'), filter={'environment_scope': var_obj.get('environment_scope')})
+        return True
 
 
-def native_python_main(this_gitlab, purge, var_list, state, module):
+def native_python_main(this_gitlab, purge, requested_variables, state, module):
 
     change = False
     return_value = dict(added=list(), updated=list(), removed=list(), untouched=list())
 
     gitlab_keys = this_gitlab.list_all_project_variables()
-    existing_variables = [x.get_id() for x in gitlab_keys]
-
-    for key in var_list:
-
-        if isinstance(var_list[key], string_types) or isinstance(var_list[key], (integer_types, float)):
-            value = var_list[key]
-            masked = False
-            protected = False
-            variable_type = 'env_var'
-            environment_scope = None
-        elif isinstance(var_list[key], dict):
-            value = var_list[key].get('value')
-            masked = var_list[key].get('masked', False)
-            protected = var_list[key].get('protected', False)
-            variable_type = var_list[key].get('variable_type', 'env_var')
-            environment_scope = var_list[key].get('environment_scope')
+    existing_variables = [x.attributes for x in gitlab_keys]
+
+    # preprocessing:filter out and enrich before compare
+    for item in existing_variables:
+        item.pop('project_id')
+
+    for item in requested_variables:
+        item['key'] = item.pop('name')
+        item['value'] = str(item.get('value'))
+        if item.get('protected') is None:
```

Comment:
this does not break backwards compatibility.  
but it might result in unexpected behaviours!

Even if we do not enrich the properties with default values, it will also result in unexpected behaviours for the users. Because when they are missing, it is always detect as a change.

I have no idea how we can save the situation...  

cc @felixfontein

Reply:
Ok, I've changed it a little so that it fits to the current behaviour.  

All what's left to make integration test pass, is the identification of `untouched` variables.